### PR TITLE
 Only apply default-members when building root manifest

### LINKED ISSUE #5932: cargo build in a workspaced crate builds workspaces' default members
I [have a virtual workspace where I override the `default-members`](https://github.com/Nemo157/embrio-rs/blob/d67a3eaf25c3503b3586c8450e0176654ca69d4e/Cargo.toml). I expected this to just set the package that builds when in the root of the virtual workspace, however when I go into one of the members sub-directories I find that it still only builds the default members:

```
~/sources/embrio-rs → cargo build --verbose
       Fresh futures-core-preview v0.3.0-alpha.3 (https://github.com/rust-lang-nursery/futures-rs#09617ac0)
       Fresh either v1.5.0
       Fresh pin-utils v0.1.0-alpha.1
       Fresh memchr v2.0.1
       Fresh futures-channel-preview v0.3.0-alpha.3 (https://github.com/rust-lang-nursery/futures-rs#09617ac0)
       Fresh futures-io-preview v0.3.0-alpha.3 (https://github.com/rust-lang-nursery/futures-rs#09617ac0)
       Fresh embrio-core v0.1.0 (file:///Users/Nemo157/sources/embrio-rs/embrio-core)
       Fresh futures-sink-preview v0.3.0-alpha.3 (https://github.com/rust-lang-nursery/futures-rs#09617ac0)
       Fresh futures-util-preview v0.3.0-alpha.3 (https://github.com/rust-lang-nursery/futures-rs#09617ac0)
       Fresh embrio-util v0.1.0 (file:///Users/Nemo157/sources/embrio-rs/embrio-util)
       Fresh embrio v0.1.0 (file:///Users/Nemo157/sources/embrio-rs/embrio)
    Finished dev [unoptimized + debuginfo] target(s) in 0.09s

~/sources/embrio-rs/ → cd embrio-native

~/sources/embrio-rs/embrio-native → cargo build --verbose
       Fresh futures-core-preview v0.3.0-alpha.3 (https://github.com/rust-lang-nursery/futures-rs#09617ac0)
       Fresh pin-utils v0.1.0-alpha.1
       Fresh either v1.5.0
       Fresh memchr v2.0.1
       Fresh futures-channel-preview v0.3.0-alpha.3 (https://github.com/rust-lang-nursery/futures-rs#09617ac0)
       Fresh futures-io-preview v0.3.0-alpha.3 (https://github.com/rust-lang-nursery/futures-rs#09617ac0)
       Fresh embrio-core v0.1.0 (file:///Users/Nemo157/sources/embrio-rs/embrio-core)
       Fresh futures-sink-preview v0.3.0-alpha.3 (https://github.com/rust-lang-nursery/futures-rs#09617ac0)
       Fresh futures-util-preview v0.3.0-alpha.3 (https://github.com/rust-lang-nursery/futures-rs#09617ac0)
       Fresh embrio-util v0.1.0 (file:///Users/Nemo157/sources/embrio-rs/embrio-util)
       Fresh embrio v0.1.0 (file:///Users/Nemo157/sources/embrio-rs/embrio)
    Finished dev [unoptimized + debuginfo] target(s) in 0.07s
```

### PR DESCRIPTION
**Pull request by Nemo157**
Fixes #5932 

The first commit includes all actual changes, the second commit is just running Asciidoctor on the already updated man-pages source, so I suggest looking at the individual commits.

### CODE REVIEW

File: src/doc/man/options-packages.adoc [hunk 1]
```diff
@@ -1,9 +1,14 @@
...
-all workspace members are selected (`--all` is implied). Otherwise, only the
-package in the current directory will be selected. The default packages may be
-overridden with the `workspace.default-members` key in the root `Cargo.toml`
-manifest.
+depend on the selected manifest file (based on the current working directory if
+`--manifest-path` is not given). If the manifest is the root of a workspace then
+the workspaces default members are selected, otherwise only the package defined
+by the manifest will be selected.
+
+The default members of a workspace can either be set explicitly with the
```
Comment (ehuss) [inline src/doc/man/options-packages.adoc]:
```suggestion
The default members of a workspace can be set explicitly with the
```

I think *either* used in this way needs two clauses within the same sentence.  It reads a little more clearly to me without it.

File: src/doc/man/options-packages.adoc [hunk 2]
```diff
@@ -1,9 +1,14 @@
...
+`workspace.default-members` key in the root manifest. If this is not set then
```
Comment (ehuss) [inline src/doc/man/options-packages.adoc]:
```suggestion
`workspace.default-members` key in the root manifest. If this is not set,
```

File: src/doc/man/options-packages.adoc [hunk 3]
```diff
@@ -1,9 +1,14 @@
...
+for a virtual workspace this will include all workspace members (equivalent to
```
Comment (ehuss) [inline src/doc/man/options-packages.adoc]:
```suggestion
a virtual workspace will include all workspace members (equivalent to
```

File: src/doc/man/options-packages.adoc [hunk 4]
```diff
@@ -1,9 +1,14 @@
...
+passing `--all`), or for a non-virtual workspace it will include only the root
```
Comment (ehuss) [inline src/doc/man/options-packages.adoc]:
```suggestion
passing `--all`), and a non-virtual workspace will include only the root
```

**Comment by ehuss** [comment]
Thanks for updating the documentation. :heart:

I'm slightly positive to this change. I want to see what others at @rust-lang/cargo think about it. The summary is:

This changes `default-members` to be ignored when in a non-root member to only build that member.

Without default-members, nothing changes:

Command | Virtual | Packages Selected
--------|---------|------------------
`cargo` in `ws/` | ✓ | All members
`cargo` in `ws/` |   | Root crate only
`cargo` in `ws/foo/` | ✓ | `foo` only
`cargo` in `ws/foo/` |   | `foo` only

With default members, this PR changes the last two lines in this table from `default-members` to "`foo` only":

Command | Virtual | Packages Selected
--------|---------|------------------
`cargo` in `ws/` | ✓ | `default-members` only
`cargo` in `ws/` |   | `default-members` only
`cargo` in `ws/foo/` | ✓ | **`default-members` only**
`cargo` in `ws/foo/` |   | **`default-members` only**

I'm a little concerned this may disrupt existing workflows. It may be a little more consistent with the non-default-members case.

I'm also curious what @SimonSapin thinks. He added it for servo, though it appears servo no longer uses it. I don't use any projects with this feature, so I don't have any experience with it.

**Comment by alexcrichton** [comment]
This seems like a reasonable change in behavior to me as well!

**Comment by Nemo157** [comment]
I've rebased your doc suggestions in.

**Comment by alexcrichton** [comment]
I'm gonna close this since it's been inactive for some time now, but it can of course be resubmitted! (just trying to help clear out the queue)

**Comment by plauche** [comment]
@Nemo157 any plans to pick this back up? I've run into the same unexpected behaviour when using `default-members`. I may restart this work if you aren't planning on continuing with it.

**Comment by Nemo157** [comment]
I've rebased the branch, and the tests still pass.

I was waiting on the feedback that @ehuss requested above, if they think that consensus on merging this can be had I'll happily open another PR for it. I'm not working on the project where I was having this issue much anymore so I didn't have the motivation to try and push for getting this merged.

**Comment by plauche** [comment]
@ehuss do you have any input on moving forward with this PR? I'd like to use the `default-members` feature, however the current behaviour is when working in member crates is not very desirable.

**Comment by ehuss** [comment]
I have posted a request for feedback at https://internals.rust-lang.org/t/cargo-change-in-workspace-default-members/10649.  If there aren't any significant objections, then I'll go ahead and approve this PR.

**Comment by ehuss** [comment]
I haven't seen any objections, so I'm going to go ahead and merge this. However, github won't let me reopen, so I will open a new PR and merge that (#7270).

**Pull request closed without merging**